File: src/main/java/org/nyxlang/interpreter/visitor/NumberVisitor.java
package org.nyxlang.interpreter.visitor;

import org.nyxlang.interpreter.IInterpreter;
import org.nyxlang.interpreter.exception.VisitorException;
import org.nyxlang.parser.ast.INode;
import org.nyxlang.parser.ast.NumberNode;

import java.math.BigDecimal;


/**
 * Interprets a number node.
 */
public class NumberVisitor implements IVisitor {

    @Override
    public boolean matches(INode node) {
        return NumberNode.class == node.getClass();
    }

    @Override
    public BigDecimal visit(IInterpreter interpreter, INode node) throws VisitorException {
        return ((NumberNode) node).getValue();
    }
}

File: src/main/kotlin/org/nyxlang/interpreter/visitor/NumberVisitor.kt
package org.nyxlang.interpreter.visitor

import org.nyxlang.interpreter.IInterpreter
import org.nyxlang.interpreter.result.numberResult
import org.nyxlang.parser.ast.INode
import org.nyxlang.parser.ast.NumberNode

/**
 * Interprets a number node.
 */
class NumberVisitor : IVisitor {
    override fun visit(interpreter: IInterpreter, node: INode) = numberResult((node as NumberNode).value)
}